core/lib/src/shutdown.rs

```diff
@@ -0,0 +1,92 @@
+use futures::{channel::mpsc, stream::StreamExt};
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// Wrapper around an mpsc channel.
+#[derive(Debug)]
+crate struct Shutdown {
+    sender: mpsc::Sender<()>,
+    crate receiver: Option<Receiver>,
+}
+
+impl Shutdown {
+    /// Create a `Shutdown`.
+    #[inline]
+    crate fn new() -> Self {
+        let (sender, receiver) = mpsc::channel(1);
+        Self {
+            sender,
+            receiver: Some(Receiver { receiver }),
+        }
+    }
+
+    /// Get a handle that can be used to call `.shutdown()`.
+    #[inline]
+    crate fn get_handle(&self) -> ShutdownHandle<'_> {
+        ShutdownHandle {
+            sender: &self.sender,
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # #![feature(proc_macro_hygiene, async_await)]
+/// # #[macro_use] extern crate rocket;
+/// #
+/// use rocket::shutdown::ShutdownHandle;
+///
+/// #[get("/shutdown")]
+/// fn shutdown(handle: &ShutdownHandle) -> &'static str {
+///     handle.shutdown();
+///     "Shutting down..."
+/// }
+///
+/// fn main() {
+///     # if false {
+///     rocket::ignite()
+///         .mount("/", routes![shutdown])
+///         .launch();
+///     # }
+/// }
+/// ```
+#[derive(Debug, Clone, Copy)]
+pub struct ShutdownHandle<'a> {
+    sender: &'a mpsc::Sender<()>,
+}
+
+impl ShutdownHandle<'_> {
+    /// Notify Rocket to shut down gracefully.
+    ///
+    /// This method should only need to be called once.
```

Comment:
"should" isn't very strong. There is no reason calling once would fail but twice would work, and calling more than once is not an error.

Reply:
"This method only needs to be called once"? Keep it straight to the point, while not even hinting at what could/does happen when called a second time.